src/resonate/scheduler.py

```diff
@@ -649,66 +816,76 @@ def _resolve_promise(
+            ):
+                runnable.coro.route_info.retry_attempt += 1
+
+                self._delay_queue.put_nowait(
+                    runnable.coro.route_info,
+                    delay=_next_retry_delay(
+                        runnable.coro.route_info.retry_policy,
+                        runnable.coro.route_info.retry_attempt,
+                    ),
+                )
+
+            else:
                 self._tracing_adapter.process_event(
                     ExecutionTerminated(
-                        promise_id=runnable.coro_and_promise.route_info.promise.promise_id,
+                        promise_id=runnable.coro.route_info.promise.promise_id,
                         tick=now(),
-                        parent_promise_id=runnable.coro_and_promise.route_info.promise.parent_promise_id(),
-                    )
-                )
-                if all_promises_are_done(
-                    runnable.coro_and_promise.route_info.promise.children_promises
-                ):
-                    self._resolve_promise(
-                        runnable.coro_and_promise.route_info.promise, final_value
-                    )
-                    self._unblock_coros_waiting_on_promise(
-                        runnable.coro_and_promise.route_info.promise
+                        parent_promise_id=runnable.coro.route_info.promise.parent_promise_id(),
                     )
-                else:
-                    runnable.coro_and_promise.final_value = yieldable_or_final_value
-                    callback_registered: bool = False
-                    for (
-                        child_promise
-                    ) in runnable.coro_and_promise.route_info.promise.children_promises:
-                        if not child_promise.done():
-                            self._add_coro_to_awaitables(
-                                child_promise, runnable.coro_and_promise
-                            )
-                            if isinstance(child_promise.action, RFI):
-                                if not callback_registered:
-                                    self._register_callback_or_resolve_ephemeral_promise(
-                                        child_promise
-                                    )
-                                callback_registered = True
-
-                        self._signal()
-            else:
-                self._delay_queue.put_nowait(
-                    runnable.coro_and_promise.route_info.next_retry_attempt(),
-                    delay=runnable.coro_and_promise.route_info.next_retry_delay(),
                 )
+                assert all_promises_are_done(
+                    runnable.coro.route_info.promise.children_promises
+                ), "All children promise must have been resolved."
+                self._resolve_promise(runnable.coro.route_info.promise, final_value)
+                self._unblock_coros_waiting_on_promise(runnable.coro.route_info.promise)
```

Comment:
Similar to the above comment, I don't think we need this code anymore, if I understand the new way of doing structured concurrency, we won't get a final value until all promises have being internally "yielded" right?

Reply:
Removed. Moving assert statement into structure concurrency